File: app/src/main/java/cf/playhi/freezeyou/FUFLauncherShortcutCreator.java
package cf.playhi.freezeyou;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.View;
import android.widget.AbsListView;
import android.widget.AdapterView;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.ProgressBar;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import cf.playhi.freezeyou.app.FreezeYouBaseActivity;
import cf.playhi.freezeyou.utils.OneKeyListUtils;
import cf.playhi.freezeyou.utils.ToastUtils;

import static cf.playhi.freezeyou.ThemeUtils.getThemeDot;
import static cf.playhi.freezeyou.ThemeUtils.getThemeSecondDot;
import static cf.playhi.freezeyou.ThemeUtils.processActionBar;
import static cf.playhi.freezeyou.ThemeUtils.processSetTheme;
import static cf.playhi.freezeyou.utils.ApplicationIconUtils.getApplicationIcon;
import static cf.playhi.freezeyou.utils.ApplicationLabelUtils.getApplicationLabel;
import static cf.playhi.freezeyou.utils.FUFUtils.realGetFrozenStatus;
import static cf.playhi.freezeyou.utils.MoreUtils.processListFilter;

public class FUFLauncherShortcutCreator extends FreezeYouBaseActivity {

    private int customThemeDisabledDot = R.drawable.shapedotblue;
    private int customThemeEnabledDot = R.drawable.shapedotblack;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        processSetTheme(this);
        super.onCreate(savedInstanceState);
        processActionBar(getSupportActionBar());

        final Intent intent = getIntent();
        String slf_n = intent.getStringExtra("slf_n");
        final boolean returnPkgName = intent.getBooleanExtra("returnPkgName", false);
        final boolean isSlfMode = slf_n != null;

        try {
            customThemeDisabledDot = getThemeDot(this);
            customThemeEnabledDot = getThemeSecondDot(this);
        } catch (Exception e) {
            e.printStackTrace();
        }

        if (isSlfMode || returnPkgName) {
            setContentView(R.layout.fuflsc_select_application);

            if (isSlfMode)
                setTitle(R.string.add);
            else
                setTitle(R.string.plsSelect);

            new Thread(new Runnable() {
                @Override
                public void run() {

                    final ListView app_listView = findViewById(R.id.fuflsc_app_list);
                    final ProgressBar progressBar = findViewById(R.id.fuflsc_progressBar);
                    final LinearLayout linearLayout = findViewById(R.id.fuflsc_linearLayout);
                    final ArrayList<Map<String, Object>> AppList = new ArrayList<>();
                    final EditText search_editText = findViewById(R.id.fuflsc_search_editText);
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            linearLayout.setVisibility(View.VISIBLE);
                            progressBar.setVisibility(View.VISIBLE);
                            app_listView.setVisibility(View.GONE);
                        }
                    });
                    app_listView.setChoiceMode(AbsListView.CHOICE_MODE_NONE);
                    final Context applicationContext = getApplicationContext();
                    PackageManager packageManager = applicationContext.getPackageManager();
                    List<ApplicationInfo> applicationInfo = packageManager.getInstalledApplications(PackageManager.GET_UNINSTALLED_PACKAGES);
                    int size = applicationInfo == null ? 0 : applicationInfo.size();
                    for (int i = 0; i < size; i++) {
                        ApplicationInfo applicationInfo1 = applicationInfo.get(i);
                        Map<String, Object> keyValuePair = processAppStatus(
                                getApplicationLabel(applicationContext, packageManager, applicationInfo1, applicationInfo1.packageName),
                                applicationInfo1.packageName,
                                applicationInfo1,
                                packageManager
                        );
                        if (keyValuePair != null) {
                            AppList.add(keyValuePair);
                        }
                    }

                    if (!AppList.isEmpty()) {
                        Collections.sort(AppList, new Comparator<Map<String, Object>>() {

                            @Override
                            public int compare(Map<String, Object> stringObjectMap, Map<String, Object> t1) {
                                return ((String) stringObjectMap.get("PackageName")).compareTo((String) t1.get("PackageName"));
                            }
                        });
                    }

                    final ReplaceableSimpleAdapter adapter =
                            new ReplaceableSimpleAdapter(
                                    FUFLauncherShortcutCreator.this,
                                    (ArrayList<Map<String, Object>>) AppList.clone(),
                                    R.layout.app_list_1,
                                    new String[]{"Img", "Name", "PackageName", "isFrozen"},
                                    new int[]{R.id.img, R.id.name, R.id.pkgName, R.id.isFrozen});//isFrozen、isAutoList传图像资源id

                    adapter.setViewBinder(new ReplaceableSimpleAdapter.ViewBinder() {
                        public boolean setViewValue(View view, Object data, String textRepresentation) {
                            if (view instanceof ImageView && data instanceof Drawable) {
                                ((ImageView) view).setImageDrawable((Drawable) data);
                                return true;
                            } else
                                return false;
                        }
                    });

                    search_editText.addTextChangedListener(new TextWatcher() {
                        @Override
                        public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

                        }

                        @Override
                        public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                            if (TextUtils.isEmpty(charSequence)) {
                                adapter.replaceAllInFormerArrayList(AppList);
                            } else {
                                adapter.replaceAllInFormerArrayList(processListFilter(charSequence, AppList));
                            }
                        }

                        @Override
                        public void afterTextChanged(Editable editable) {

                        }
                    });

                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            progressBar.setVisibility(View.GONE);
                            linearLayout.setVisibility(View.GONE);
                            app_listView.setAdapter(adapter);
                            app_listView.setTextFilterEnabled(true);
                            app_listView.setVisibility(View.VISIBLE);
                        }
                    });

                    app_listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
                        @Override
                        public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
                            HashMap<String, Object> map = (HashMap<String, Object>) app_listView.getItemAtPosition(i);
                            final String name = (String) map.get("Name");
                            final String pkgName = (String) map.get("PackageName");
                            if (isSlfMode) {
                                SharedPreferences sp = getSharedPreferences(getIntent().getStringExtra("slf_n"), MODE_PRIVATE);
                                if (!OneKeyListUtils.existsInOneKeyList(sp.getString("pkgS", ""), pkgName)) {
                                    sp.edit().putString("pkgS", sp.getString("pkgS", "") + pkgName + ",").apply();
                                    ToastUtils.showToast(FUFLauncherShortcutCreator.this, R.string.added);
                                } else {
                                    ToastUtils.showToast(FUFLauncherShortcutCreator.this, R.string.alreadyExist);
                                }
                                setResult(RESULT_OK);
                            } else {// if (returnPkgName)s
                                setResult(RESULT_OK, new Intent()
                                        .putExtra("pkgName", pkgName)
                                        .putExtra("name", name)
                                        .putExtra("id", "FreezeYou! " + pkgName));
                                finish();
                            }
                        }
                    });
                }
            }).start();

        } else {
            finish();
        }

    }

    private Map<String, Object> processAppStatus(String name, String packageName, ApplicationInfo applicationInfo, PackageManager packageManager) {
        if (!("android".equals(packageName) || "cf.playhi.freezeyou".equals(packageName))) {
            Map<String, Object> keyValuePair = new HashMap<>();
            keyValuePair.put("Img", getApplicationIcon(FUFLauncherShortcutCreator.this, packageName, applicationInfo, true));
            keyValuePair.put("Name", name);
            processFrozenStatus(keyValuePair, packageName, packageManager);
            keyValuePair.put("PackageName", packageName);
            return keyValuePair;
        }
        return null;
    }

//    /**
//     * @param packageName 应用包名
//     * @return true 则已冻结
//     */
//    private boolean realGetFrozenStatus(String packageName, PackageManager pm) {
//        return (checkRootFrozen(FUFLauncherShortcutCreator.this, packageName, pm) || checkMRootFrozen(FUFLauncherShortcutCreator.this, packageName));
//    }

    private void processFrozenStatus(Map<String, Object> keyValuePair, String packageName, PackageManager packageManager) {
        keyValuePair.put("isFrozen", getFrozenStatus(packageName, packageManager));
    }

    /**
     * @param packageName 应用包名
     * @return 资源 Id
     */
    private int getFrozenStatus(String packageName, PackageManager packageManager) {
        return realGetFrozenStatus(FUFLauncherShortcutCreator.this, packageName, packageManager)
                ? customThemeDisabledDot : customThemeEnabledDot;
    }

}

File: app/src/main/java/cf/playhi/freezeyou/ui/FUFLauncherShortcutCreator.kt
package cf.playhi.freezeyou.ui

import android.content.Intent
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.graphics.drawable.Drawable
import android.os.Build
import android.os.Bundle
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.res.stringResource
import androidx.core.os.ConfigurationCompat
import cf.playhi.freezeyou.R
import cf.playhi.freezeyou.app.FreezeYouBaseActivity
import cf.playhi.freezeyou.ui.compose.AppListItem
import cf.playhi.freezeyou.ui.compose.FreezeYouTheme
import cf.playhi.freezeyou.utils.ApplicationIconUtils.getApplicationIcon
import cf.playhi.freezeyou.utils.ApplicationLabelUtils.getApplicationLabel
import cf.playhi.freezeyou.utils.FUFUtils.realGetFrozenStatus
import cf.playhi.freezeyou.utils.OneKeyListUtils.existsInOneKeyList
import cf.playhi.freezeyou.utils.ThemeUtils
import cf.playhi.freezeyou.utils.ToastUtils.showToast
import java.util.Locale

class FUFLauncherShortcutCreator : FreezeYouBaseActivity() {
    private data class AppEntry(
        val icon: Drawable?, val name: String, val packageName: String, val statusIcon: Int
    )

    private var customThemeDisabledDot = R.drawable.shapedotblue
    private var customThemeEnabledDot = R.drawable.shapedotblack
    private var applications by mutableStateOf<List<AppEntry>>(emptyList())
    private var loading by mutableStateOf(true)
    private var search by mutableStateOf("")

    override fun onCreate(savedInstanceState: Bundle?) {
        ThemeUtils.processSetTheme(this)
        super.onCreate(savedInstanceState)
        ThemeUtils.processActionBar(supportActionBar)
        val folderName = intent.getStringExtra("slf_n")
        val returnPackageName = intent.getBooleanExtra("returnPkgName", false)
        if (folderName == null && !returnPackageName) {
            finish()
            return
        }
        setTitle(if (folderName != null) R.string.add else R.string.plsSelect)
        try {
            customThemeDisabledDot = ThemeUtils.getThemeDot(this)
            customThemeEnabledDot = ThemeUtils.getThemeSecondDot(this)
        } catch (exception: Exception) {
            exception.printStackTrace()
        }
        setContent {
            FreezeYouTheme {
                val locale = ConfigurationCompat.getLocales(LocalConfiguration.current)[0]
                    ?: Locale.ROOT
                val query = search.lowercase(locale)
                val displayed = if (query.isEmpty()) applications else applications.filter {
                    it.name.lowercase(locale).contains(query) ||
                        it.packageName.lowercase(locale).contains(query)
                }
                Column(Modifier.fillMaxSize()) {
                    OutlinedTextField(
                        value = search,
                        onValueChange = { search = it },
                        label = { Text(stringResource(R.string.search)) },
                        singleLine = true,
                        modifier = Modifier.fillMaxWidth()
                    )
                    if (loading) LinearProgressIndicator(Modifier.fillMaxWidth())
                    LazyColumn(Modifier.fillMaxSize()) {
                        items(displayed, key = { it.packageName }) { app ->
                            AppListItem(
                                icon = app.icon,
                                title = app.name,
                                subtitle = app.packageName,
                                statusIcon = app.statusIcon
                            ) { selectApp(app, folderName) }
                        }
                    }
                }
            }
        }
        loadApplications()
    }

    private fun loadApplications() {
        Thread {
            val packageManager = applicationContext.packageManager
            val matchUninstalled = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                PackageManager.MATCH_UNINSTALLED_PACKAGES
            } else {
                @Suppress("DEPRECATION")
                PackageManager.GET_UNINSTALLED_PACKAGES
            }
            val result = packageManager
                .getInstalledApplications(matchUninstalled)
                .mapNotNull { processAppStatus(it, packageManager) }
                .sortedBy { it.packageName }
            runOnUiThread {
                applications = result
                loading = false
            }
        }.start()
    }

    private fun selectApp(app: AppEntry, folderName: String?) {
        if (folderName != null) {
            val preferences = getSharedPreferences(folderName, MODE_PRIVATE)
            if (!existsInOneKeyList(preferences.getString("pkgS", ""), app.packageName)) {
                preferences.edit()
                    .putString("pkgS", preferences.getString("pkgS", "") + app.packageName + ",")
                    .apply()
                showToast(this, R.string.added)
            } else {
                showToast(this, R.string.alreadyExist)
            }
            setResult(RESULT_OK)
        } else {
            setResult(
                RESULT_OK,
                Intent()
                    .putExtra("pkgName", app.packageName)
                    .putExtra("name", app.name)
                    .putExtra("id", "FreezeYou! ${app.packageName}")
            )
            finish()
        }
    }

    private fun processAppStatus(
        applicationInfo: ApplicationInfo,
        packageManager: PackageManager
    ): AppEntry? {
        val packageName = applicationInfo.packageName
        if (packageName == "android" || packageName == "cf.playhi.freezeyou") return null
        return AppEntry(
            icon = getApplicationIcon(this, packageName, applicationInfo, true),
            name = getApplicationLabel(this, packageManager, applicationInfo, packageName),
            packageName = packageName,
            statusIcon = if (realGetFrozenStatus(this, packageName, packageManager)) {
                customThemeDisabledDot
            } else {
                customThemeEnabledDot
            }
        )
    }
}
